ethcore/sync/src/blocks.rs

```diff
@@ -278,30 +358,33 @@ impl BlockCollection {
 			while let Some(h) = head {
 				head = self.parents.get(&h).cloned();
 				if let Some(head) = head {
-					match self.blocks.get(&head) {
-						Some(block) if block.body.is_some() && (!self.need_receipts || block.receipts.is_some()) => {
-							blocks.push(block);
-							hashes.push(head);
-							self.head = Some(head);
-						}
-						_ => break,
+					match self.blocks.remove(&head) {
+						Some(block) => {
+							if block.body.is_some() && (!self.need_receipts || block.receipts.is_some()) {
+								blocks.push(block);
+								hashes.push(head);
+								self.head = Some(head);
+							} else {
+								self.blocks.insert(head, block);
```

Comment:
So previously this would have removed an already downloaded header? So adding it back is just preventing it from being downloaded again?

Reply:
so the logic has not changed here. previously we called `self.blocks.get(...)` (line 281) and then `self.blocks.remove(...)` line 303. Now we are instantly removing `self.blocks.remove(...)` (line 361), but we need to readd blocks that shouldn't be removed (this line)

Reply:
it lets us avoid unnecessary searching in the hashmap, and unnecessary copying of the blocks (few lines below)